zerver/tests/test_ldap.py

```diff
@@ -0,0 +1,109 @@
+"""
+
+class DjangoToLDAPUsernameTests(SlapdTestCase):
+    def setUp(self) -> None:
+        super().setUp()
+        self.backend = ZulipLDAPAuthBackend()
+
+    def test_django_to_ldap_username_with_append_domain(self) -> None:
+        with self.settings(LDAP_APPEND_DOMAIN="zulip.com"):
+            self.assertEqual(self.backend.django_to_ldap_username("hamlet"), "hamlet")
+            self.assertEqual(self.backend.django_to_ldap_username("hamlet@zulip.com"), "hamlet")
+            with self.assertRaises(ZulipLDAPExceptionOutsideDomain):
+                self.backend.django_to_ldap_username("hamlet@example.com")
+
+    def test_django_to_ldap_username_without_email_search(self) -> None:
+        with self.settings(AUTH_LDAP_REVERSE_EMAIL_SEARCH=None):
+            self.assertEqual(self.backend.django_to_ldap_username("hamlet"), "hamlet")
+            self.assertEqual(self.backend.django_to_ldap_username("hamlet@zulip.com"), "hamlet@zulip.com")
+            self.assertEqual(self.backend.django_to_ldap_username("hamlet@example.com"), "hamlet@example.com")
+
+    def test_django_to_ldap_username_with_email_search(self) -> None:
+        with self.settings():
+            self.assertEqual(self.backend.django_to_ldap_username("hamlet"), "hamlet")
+            self.assertEqual(self.backend.django_to_ldap_username("hamlet@zulip.com"), "hamlet")
+            self.assertEqual(self.backend.django_to_ldap_username("no_such_email@example.com"),
+                             "no_such_email@example.com")
+            # aaron has uid=letham in our test directory:
+            self.assertEqual(self.backend.django_to_ldap_username("aaron@zulip.com"), "letham")
+
+            with mock.patch("zproject.backends.logging.warning") as mock_warn:
+                self.assertEqual(
+                    self.backend.django_to_ldap_username("shared_email@zulip.com"),
+                    "shared_email@zulip.com"
+                )
+                mock_warn.assert_called_with("Multiple users with email shared_email@zulip.com found in LDAP.")
+
+        # Configure email search for emails in the uid attribute:
+        with self.settings(AUTH_LDAP_REVERSE_EMAIL_SEARCH=LDAPSearch("ou=users,dc=zulip,dc=com",
+                                                                     ldap.SCOPE_SUBTREE,
+                                                                     "(uid=%(email)s)")):
+            self.assertEqual(self.backend.django_to_ldap_username("newuser_email_as_uid@zulip.com"),
+                             "newuser_email_as_uid@zulip.com")
+
+    @override_settings(AUTHENTICATION_BACKENDS=('zproject.backends.EmailAuthBackend',
+                                                'zproject.backends.ZulipLDAPAuthBackend',))
+    def test_authenticate_to_ldap_via_email(self) -> None:
+        """
+        With django_to_ldap_username now able to translate an email to ldap username,
+        it should be possible to authenticate through user_profile.email.
```

Comment:
Code docstrings shouldn't say things like "now able to translate ...", because after these changes are merged, the historical fact that this feature was added later won't be relevant to anyone.  We should instead talk about "With AUTH_LDAP_REVERSE_EMAIL_SEARCH configured, ...`

Reply:
Changed to
```
        """
        With AUTH_LDAP_REVERSE_EMAIL_SEARCH configured, django_to_ldap_username
        should be able to translate an email to ldap username,
        and thus it should be possible to authenticate through user_profile.email.
        """
```